bundle/edu.gemini.ags.client.impl/src/main/scala/edu/gemini/ags/client/impl/AgsHttpClient.scala

```diff
@@ -39,12 +47,18 @@ case class AgsHttpClient(host: String, port: Int) extends AgsClient {
   private def estimateNow(url: URL): AgsResult = {
     try {
       AgsHttpClient.Log.info(s"AGS Query to $url")
-      val conn = url.openConnection().asInstanceOf[HttpsURLConnection]
-      conn.setHostnameVerifier(hostnameVerifier)
-      conn.setSSLSocketFactory(GemSslSocketFactory.get)
-      conn.setReadTimeout(timeout)
-      Charset.set(conn)
-      Response(conn).result
+      Option(lRUCache.get(url.toString) : AgsResult) | {
+        val conn = url.openConnection().asInstanceOf[HttpsURLConnection]
+        conn.setHostnameVerifier(hostnameVerifier)
+        conn.setSSLSocketFactory(GemSslSocketFactory.get)
+        conn.setReadTimeout(timeout)
+        Charset.set(conn)
+
+        Response(conn).result match {
+          case s@AgsResult.Success(_) => lRUCache.put(url.toString, s)
```

Comment:
This looks like a bug (?) because a Java `HashMap` returns "the previous value associated with key, or `null` if there was no mapping for key".

Also, there's a side-effect here in updating the cache.  I think it would be a tiny bit clearer to separate the side-effect from the computation.  (Never mind that the computation isn't referentially transparent anyway!). That is, store the result of the computation in a value.  Do the cache updating on success afterwords (even if that means often replacing a success value with the same value you just pulled out of the cache), then just use the value as the result.